File: day08/day08.clj
;; puzzle input : Picture of the BIOS password
;; format : Space Image Format
;;;;; series of digits that each represent the color of a single pixel
;;;;; digits fill each row of the image left-to-right
;;;;; Note: partition ?

;; image = series of identically-sized layers
(comment
  (partition 2 (partition 3 "123456789012")))

;; width 25 height 6
;; fewest 0 digits
;; the number of 1 digits multiplied by the number of 2 digits

(defn parse-input [input]
  (map #(Integer/parseInt %)
       (clojure.string/split input #"")))

(->> (slurp "day08/input")
     parse-input
     (partition (* 25 6))
     (map frequencies)
     (sort-by #(get % 0 0))
     first
     ((fn [freqs] (* (get freqs 1) (get freqs 2)))))

;; 0 is black, 1 is white, and 2 is transparent
;; first layer in front and the last layer in back
;; 2 2 0 1 -> 0
(def ->char { 0 " " 1 "X"})
(defn solve2 [width height input]
  (let [digits (parse-input input)
        layer-size (* width height)
        n-layers (quot (count digits) layer-size)]
    (->> digits
         (partition layer-size)
         (apply interleave)
         (partition n-layers)
         (map #(first (drop-while #{2} %)))
         (partition width))))

(clojure.pprint/pprint (solve2 2 2 "0222112222120000"))

(clojure.pprint/pprint (solve2 25 6 (slurp "day08/input")))

File: day08/day08.c
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[])
{
    int digit;
    int digits[20000];
    int count = 0;
    while (scanf("%1d", &digit) != EOF)
    {
        digits[count] = digit;
        count++;
    }

    int width = atoi(argv[1]);
    int height = atoi(argv[2]);
    int layerSize = width * height;
    int layers = count / layerSize;
    for (int j = 0; j < height; j++)
    {
        for (int i = 0; i < width; i++)
        {
            int pixel;
            for (int z = 0; (pixel = digits[i + (width * j) + (layerSize * z)]) == 2; z++)
                ;
            printf("%s", pixel ? "X" : " ");
        }
        printf("\n");
    }

    return 0;
}
